### ISSUE TITLE
Fix for Cwd.pm (chroot)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2728]
Migrated from [rt.perl.org#4467](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=4467) (status was 'resolved')

Searchable as RT4467$

**Comment by p5pRT** [comment]
### From Eric.Coe@oracle.com

Hi\.

Attached is a patch for Cwd\.pm to fix a problem occurring in the 'Cwd&#8203;::cwd' call
in some chroot\(2\)'ed environments where the 'pwd' command is not available\.

The problem occured in the virtual server Linux environment at
http&#8203;://aitcom\.net \(server 2 \- reseller\)\, and only occurs when the script is
running as a CGI script \(with the altered root path\)\.  When running the script
from the command line of a telnet session \(which is not chroot'ed\)\, everything
worked properly\.

In the process of tracing this bug\, I tried coping the global pwd to a place
where the CGI script can use it\. This resulted in the pwd program failing with a
SIGSEGV\.  \(Which maybe explains why it wasn't made available in the first
place\.\)

The do\-it\-in\-perl variant function in the same module\, 'Cwd&#8203;::getcwd'\, works fine
in both cases; therfore I used it as a substitute when 'pwd' can't be found in
$PATH

I hope you find this fix useful\.
Thanks\.

Eric Coe

\(P\.S\. Licence of fix is the same as the rest of perl\, blah\, blah\)\.
\-\- 
\#\!/usr/bin/perl \-i4lfirm\{8\+&#8203;:2cqsynlz&#8203;:5%@&#8203;0&#8203;:3Siwr&#8203;:1Kwvx
print 'Eric E\. Coe\, JAPH \(and Oracle Principle Engineer\)
Work&#8203;:  914\.773\.8017   Eric\.Coe@&#8203;oracle\.com AIM&#8203;: "Eric Coe US"
Pager&#8203;: 800\.946\.4646\#1700744 P\-mail&#8203;: 1700744@&#8203;archwireless\.net
Home&#8203;:  \(973\) 663\-0840 ecoe@&#8203;reportweb\.com  
BTW\, the meaning of JAPH is&#8203;:'\,"\\n@&#8203;\{\[map\{\($i\,@&#8203;a\)=unpack
'ac\*'\,$\_;pack'c\*'\,map\{$\_\-$i\+\+\}@&#8203;a\}sort split/&#8203;:/\,$^I\]\}\\n"
\_\_END\_\_
As always\, the opinions expressed are solely my own\,
and not my employer's \(though they should be\!\)

**Comment by p5pRT** [comment]
### From Eric.Coe@oracle.com

<details><summary>cwdfix.diff</summary>

```diff
*** Cwd.pm.orig	Wed Oct 18 02:50:21 2000
--- Cwd.pm	Wed Oct 18 02:50:21 2000
***************
*** 86,92 ****
  # Since some ports may predefine cwd internally (e.g., NT)
  # we take care not to override an existing definition for cwd().
  
! *cwd = \&_backtick_pwd unless defined &cwd;
  
  
  # By Brandon S. Allbery
--- 86,100 ----
  # Since some ports may predefine cwd internally (e.g., NT)
  # we take care not to override an existing definition for cwd().
  
! unless(defined &cwd) {
!     # The pwd command is not available in some chroot(2)'ed environments
!     if(grep { -x "$_/pwd" } split(':', $ENV{PATH})) {
! 	*cwd = \&_backtick_pwd;
!     }
!     else {
! 	*cwd = \&getcwd;
!     }
! }
  
  
  # By Brandon S. Allbery

```
</details>
